services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathSpec.groovy

```diff
@@ -0,0 +1,942 @@
+        newCurrentPath == currentProtectedPath
+        newCurrentProtectedPath != currentProtectedPath
+        newCurrentProtectedPath == currentPath
+
+        currentLastUpdate < northbound.getFlow(flow.id).lastUpdated
+
+        and: "All rules for main and protected paths are updated"
+        Wrappers.wait(WAIT_OFFSET) { flowHelper.verifyRulesOnProtectedFlow(flow.id) }
+
+        and: "Cleanup: revert system to original state"
+        flowHelper.deleteFlow(flow.id)
+    }
+
+    def "Unable to swap a flow in case allocate_protected_path=false"() {
+        given: "Two active neighboring switches"
+        def isls = topology.getIslsForActiveSwitches()
+        def (srcSwitch, dstSwitch) = [isls.first().srcSwitch, isls.first().dstSwitch]
+
+        and: "A flow without protected path"
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flow.allocateProtectedPath = false
+        flowHelper.addFlow(flow)
+        !northbound.getFlowPath(flow.id).protectedPath
+
+        when: "Try to swap flow without protected path"
+        northbound.swapFlowPath(flow.id)
+
+        then: "Human readable error is returned"
+        def exc = thrown(HttpClientErrorException)
+        exc.rawStatusCode == 400
+        exc.responseBodyAsString.to(MessageError).errorMessage ==
+                "Could not swap paths: Flow $flow.id doesn't have protected path"
+
+        and: "Cleanup: revert system to original state"
+        flowHelper.deleteFlow(flow.id)
+    }
+
+    def "Unable to swap a non-existent flow"() {
+        when: "Try to swap path on a non-existent flow"
+        northbound.swapFlowPath(NON_EXISTENT_FLOW_ID)
+
+        then: "Human readable error is returned"
+        def exc = thrown(HttpClientErrorException)
+        exc.rawStatusCode == 404
+        exc.responseBodyAsString.to(MessageError).errorMessage ==
+                "Could not swap paths: Flow $NON_EXISTENT_FLOW_ID not found"
+
+    }
+
+    def "Unable to swap an inactive flow"() {
+        given: "Two active neighboring switches with two not overlapping paths at least"
+        def switches = topology.getActiveSwitches()
+        def (Switch srcSwitch, Switch dstSwitch) = [switches, switches].combinations()
+                .findAll { src, dst -> src != dst }.find { Switch src, Switch dst ->
+            database.getPaths(src.dpId, dst.dpId)*.path.collect {
+                pathHelper.getInvolvedIsls(it)
+            }.unique { a, b -> a.intersect(b) ? 0 : 1 }.size() >= 2
+        } ?: assumeTrue("No suiting switches found", false)
+
+        and: "A flow with protected path"
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flow.allocateProtectedPath = true
+        flowHelper.addFlow(flow)
+
+        and: "All alternative paths are unavailable (bring ports down on the source switch)"
+        List<PathNode> broughtDownPorts = []
+        database.getPaths(srcSwitch.dpId, dstSwitch.dpId)*.path.findAll { it != flow }.unique {
```

Comment:
`it != flow`
You try to compare path and flow objects.

Reply:
fixed